src/mesh/refinement_cc_in_one.hpp

```diff
@@ -36,6 +37,170 @@ TaskStatus RestrictPhysicalBounds(MeshData<Real> *md);
+  return (info.allocated && info.refinement_op == op);
+}
+
+template <int DIM>
+KOKKOS_FORCEINLINE_FUNCTION void
+RestrictCellAverage(const int l, const int m, const int n, const int ck, const int cj,
+                    const int ci, const IndexRange &ckb, const IndexRange &cjb,
+                    const IndexRange &cib, const IndexRange &kb, const IndexRange &jb,
+                    const IndexRange &ib, const Coordinates_t &coords,
+                    const ParArray6D<Real> &coarse, const ParArray6D<Real> &fine) {
+  const int i = (ci - cib.s) * 2 + ib.s;
+  int j = jb.s;
+  if (DIM > 1) {
+    j = (cj - cjb.s) * 2 + jb.s;
+  }
+  int k = kb.s;
+  if (DIM > 2) {
+    k = (ck - ckb.s) * 2 + kb.s;
+  }
+  // JMM: If dimensionality is wrong, accesses are out of bounds. Only
+  // access cells if dimensionality is correct.
+  const Real vol000 = coords.Volume(k, j, i);
+  const Real vol001 = coords.Volume(k, j, i + 1);
+  const Real fine000 = fine(l, m, n, k, j, i);
+  const Real fine001 = fine(l, m, n, k, j, i + 1);
+  Real vol010, vol011, vol100, vol101, vol110, vol111;
+  Real fine010, fine011, fine100, fine101, fine110, fine111;
+  vol010 = vol011 = vol100 = vol101 = vol110 = vol111 = 0;
+  fine010 = fine011 = fine100 = fine101 = fine110 = fine111 = 0;
+  if (DIM > 1) {
+    vol010 = coords.Volume(k, j + 1, i);
+    vol011 = coords.Volume(k, j + 1, i + 1);
+    fine010 = fine(l, m, n, k, j + 1, i);
+    fine011 = fine(l, m, n, k, j + 1, i + 1);
+  }
+  if (DIM > 2) {
+    vol100 = coords.Volume(k + 1, j, i);
+    vol101 = coords.Volume(k + 1, j, i + 1);
+    vol110 = coords.Volume(k + 1, j + 1, i);
+    vol111 = coords.Volume(k + 1, j + 1, i + 1);
+    fine100 = fine(l, m, n, k + 1, j, i);
+    fine101 = fine(l, m, n, k + 1, j, i + 1);
+    fine110 = fine(l, m, n, k + 1, j + 1, i);
+    fine111 = fine(l, m, n, k + 1, j + 1, i + 1);
+  }
+  // KGF: add the off-centered quantities first to preserve FP
+  // symmetry
+  const Real tvol =
+      ((vol000 + vol010) + (vol001 + vol011)) + ((vol100 + vol110) + (vol101 + vol111));
+  coarse(l, m, n, ck, cj, ci) =
+      (((fine000 * vol000 + fine010 * vol010) + (fine001 * vol001 + fine011 * vol011)) +
+       ((fine100 * vol100 + fine110 * vol110) + (fine101 * vol101 + fine111 * vol111))) /
+      tvol;
+}
+
+KOKKOS_FORCEINLINE_FUNCTION
+Real GradMinMod(const Real ccval, const Real fm, const Real fp, const Real dxm,
+                const Real dxp) {
+  const Real gxm = (ccval - fm) / dxm;
+  const Real gxp = (fp - ccval) / dxp;
+  return 0.5 * (SIGN(gxm) + SIGN(gxp)) * std::min(std::abs(gxm), std::abs(gxp));
+}
+
+// TODO(JMM): this could be simplified if grid spacing was always uniform
+template <int DIM>
+KOKKOS_FORCEINLINE_FUNCTION Real GetX(const Coordinates_t &coords, int i) {
+  if (DIM == 1) return coords.x1v(i);
+  if (DIM == 2) return coords.x2v(i);
+  if (DIM == 3) return coords.x3v(i);
+  PARTHENON_FAIL("Unknown dimension");
+}
```

Comment:
How about adding a `// TODO(cpp17) make this constexpr` ?
Or even implement the different types explicitly by hand (shouldn't be many more lines).